### ISSUE TITLE
is it possible to migrate from gogs to gogs?

### ISSUE THREAD

**Issue by atsepkov** [valeriangalliat/gogs-migrate#4]
My use case is a staging server that needs to sync to production every couple weeks. Both are powered by Gogs, and I would like a clean automated migration mechanism to accomplish this. Would this project be a good starting point?

**Comment by valeriangalliat** [comment]
This is definitely achievable with gogs-migrate if you don't care about syncing issues and pull requests and deleted repos. And I'll happily point you in the right direction to contribute this feature if you're willing to implement it.

But I think the best solution since you have access to both servers and both are running Gogs would rather be to do the sync at a lower level; you could sync the databases (e.g: `pg_dump` on the prod and `pg_restore` on the staging if you use PostgreSQL, copying the SQLite file if you're using SQLite and so on) in order to keep all PR, issues, users and settings, and make a shell script to mirror the Git repos themselves (would SSH in the prod and list the repos, loop through them and if it doesn't exist, `git clone --mirror` it from prod otherwise just `git remote update`, and optionally at the end, would garbage collect the repos that are on the staging that are not anymore on the prod).

I didn't test this but that's rather what I would try to do in order to get a perfect sync. What do you think?

**Comment by atsepkov** [comment]
Makes sense, that was going to be my original approach, but I wanted a bit more granularity - as in migrate some users and not others, and copying sqlite file wouldn't allow that.

**Comment by valeriangalliat** [comment]
Right. Keep in mind that gogs-migrate also just *creates* repos on the Gogs instance, optionally mirrors to tell Gogs to sync it every hour or so, but if you don't use a mirror repo, it won't sync the repos themselves. So if your staging instance already contains repos with the same names, it's not gonna do anything.

If you don't use mirror repos you'll have to write a script to force pull the production repos of the users of your selection in the staging Gogs. If you use mirror repos, I believe you won't be allowed to push on them which maybe defeats the purpose of a staging Gogs environment...

So having all that in mind I'm not sure gogs-migrate is the best tool for what you want to do. I think you're better off making your own script for that purpose.

But I also believe a staging environment should mirror production as closely as possible and the cleanest way to achieve this would be by syncing the database itself every once in a while and either `rsync`ing the whole repos directory or making a script to clone/update all the repos. And if you want to have only a custom selection of users in your staging environment, you could run a couple SQL queries to delete all users that are not in that list.
